File: sdk/php/src/BdiClient.php
<?php

namespace Panteao;

class BdiClient {
    const VERSION = '1.1.17';

    private static function downloadEngine(string $binPath): void {
        $isWin = (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN');
        $isMac = (PHP_OS === 'Darwin');
        $osName = $isWin ? 'win32' : ($isMac ? 'darwin' : 'linux');
        $arch = (php_uname('m') === 'aarch64' || php_uname('m') === 'arm64') ? 'arm64' : 'x64';
        
        $pkgName = "panteao-engine-$osName-$arch";
        $version = self::VERSION;
        $url = "https://registry.npmjs.org/$pkgName/-/$pkgName-$version.tgz";
        
        echo "\033[36m[Panteao]\033[0m Downloading native engine for $osName-$arch (v$version)...\n";
        
        $tgzData = @file_get_contents($url);
        if ($tgzData === false) throw new \Exception("Failed to download engine from $url");
        
        $tmpTgz = tempnam(sys_get_temp_dir(), 'engine') . '.tgz';
        file_put_contents($tmpTgz, $tgzData);
        
        try {
            $phar = new \PharData($tmpTgz);
            $extracted = false;
            foreach (new \RecursiveIteratorIterator($phar) as $file) {
                $filename = $file->getFilename();
                if ($filename === 'panteao-engine' || $filename === 'panteao-engine.exe') {
                    $dir = dirname($binPath);
                    if (!is_dir($dir)) mkdir($dir, 0777, true);
                    copy($file->getPathname(), $binPath);
                    chmod($binPath, 0755);
                    $extracted = true;
                    break;
                }
            }
            if (!$extracted) throw new \Exception("Binary not found in tarball");
        } finally {
            unlink($tmpTgz);
        }
    }

    private $socket;
    private $handlers = [];
    private $process;

    private static function getFreePort(): int {
        $server = stream_socket_server("tcp://127.0.0.1:0", $errno, $errstr);
        if (!$server) return 44444;
        $name = stream_socket_get_name($server, false);
        fclose($server);
        if ($name) {
            $parts = explode(':', $name);
            return (int)end($parts);
        }
        return 44444;
    }

    private static function findBinary(): string {
        $isWin = (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN');
        $binName = $isWin ? 'panteao-engine.exe' : 'panteao-engine';
        
        $currentDir = __DIR__;
        $cand1 = $currentDir . '/' . $binName;
        if (file_exists($cand1)) return $cand1;
        $cand2 = $currentDir . '/bin/' . $binName;
        if (file_exists($cand2)) return $cand2;
        
        $cwd = getcwd();
        $cand3 = $cwd . '/' . $binName;
        if (file_exists($cand3)) return $cand3;
        $cand4 = $cwd . '/bin/' . $binName;
        if (file_exists($cand4)) return $cand4;
        
        return $binName;
    }

    public function __construct(string $host = '127.0.0.1', int $port = 0, ?string $project = null) {
        $actualHost = empty($host) ? '127.0.0.1' : $host;
        if ($project !== null) {
            if ($port === 0) {
                $port = self::getFreePort();
            }
            $bin = self::findBinary();
            if (!file_exists($bin)) {
                $currentDir = __DIR__;
                $bin = $currentDir . '/' . (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN' ? 'panteao-engine.exe' : 'panteao-engine');
                self::downloadEngine($bin);
            }
            $descriptorspec = [
                0 => ["pipe", "r"],
                1 => ["file", "php://stdout", "a"],
                2 => ["file", "php://stderr", "a"]
            ];
            if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
                $descriptorspec[1] = ["file", "NUL", "a"];
                $descriptorspec[2] = ["file", "NUL", "a"];
            }
            $this->process = proc_open([$bin, $project, '--port', (string)$port], $descriptorspec, $pipes);
            usleep(800000);
        } else if ($port === 0) {
            $port = 44444;
        }

        $this->socket = fsockopen($actualHost, $port, $errno, $errstr, 5);
        if (!$this->socket) {
            if ($this->process) proc_terminate($this->process);
            throw new \Exception("Could not connect to Panteao: $errstr");
        }

        while (!feof($this->socket)) {
            $line = fgets($this->socket);
            if ($line === false) {
                if ($this->process) proc_terminate($this->process);
                throw new \Exception("Connection closed during handshake");
            }
            if (str_contains($line, '"type":"mas_ready"')) {
                break;
            }
        }
    }

    public function __destruct() {
        $this->close();
    }

        public function sendMsg(string $performative, string $sender, string $receiver, string $content): void {
        $msg = json_encode(['type' => 'message', 'performative' => $performative, 'sender' => $sender, 'receiver' => $receiver, 'content' => $content]) . "\n";
        fwrite($this->socket, $msg);
    }

    public function sendPerception(string $action, string $perception): void {
        $payload = json_encode([
            'type' => 'perception',
            'action' => $action,
            'perception' => $perception
        ]) . "\n";
        fwrite($this->socket, $payload);
    }

    public function registerAction(string $actionName, callable $callback): void {
        $this->handlers[$actionName] = $callback;
    }

    public function processActions(float $timeoutSeconds = 5.0): void {
        stream_set_timeout($this->socket, (int)$timeoutSeconds, (int)(($timeoutSeconds - (int)$timeoutSeconds) * 1000000));
        while (!feof($this->socket)) {
            $line = fgets($this->socket);
            if ($line === false) {
                break;
            }
            $line = trim($line);
            if (empty($line)) continue;
            
            $msg = json_decode($line, true);
            if ($msg && isset($msg['type']) && $msg['type'] === 'action') {
                $rawAction = $msg['action'];
                $id = $msg['id'];
                $parsed = $this->parseAction($rawAction);
                
                $handler = $this->handlers[$parsed['name']] ?? null;
                if ($handler) {
                    $respond = function(bool $success) use ($id) {
                        $this->sendActionResult($id, $success);
                    };
                    $handler($parsed['args'], $respond);
                } else {
                    $this->sendActionResult($id, true);
                }
            }
        }
    }

    private function parseAction(string $actionStr): array {
        $parenIdx = strpos($actionStr, '(');
        if ($parenIdx === false) {
            return ['name' => trim($actionStr), 'args' => []];
        }
        $name = trim(substr($actionStr, 0, $parenIdx));
        $argsStr = substr($actionStr, $parenIdx + 1, strrpos($actionStr, ')') - $parenIdx - 1);
        
        $args = [];
        $current = '';
        $insideQuotes = false;
        $depthBrackets = 0;
        $depthParens = 0;
        for ($i = 0; $i < strlen($argsStr); $i++) {
            $char = $argsStr[$i];
            if ($char === '"') {
                $insideQuotes = !$insideQuotes;
                $current .= $char;
            } else if (!$insideQuotes && $char === '[') {
                $depthBrackets++;
                $current .= $char;
            } else if (!$insideQuotes && $char === ']') {
                $depthBrackets--;
                $current .= $char;
            } else if (!$insideQuotes && $char === '(') {
                $depthParens++;
                $current .= $char;
            } else if (!$insideQuotes && $char === ')') {
                $depthParens--;
                $current .= $char;
            } else if ($char === ',' && !$insideQuotes && $depthBrackets === 0 && $depthParens === 0) {
                $args[] = $this->cleanArg($current);
                $current = '';
            } else {
                $current .= $char;
            }
        }
        if (strlen(trim($current)) > 0) {
            $args[] = $this->cleanArg($current);
        }
        return ['name' => $name, 'args' => $args];
    }

    private function cleanArg(string $arg): string {
        $s = trim($arg);
        if (str_starts_with($s, '"') && str_ends_with($s, '"') && strlen($s) >= 2) {
            return substr($s, 1, -1);
        }
        return $s;
    }

    private function sendActionResult(string $id, bool $success): void {
        $payload = json_encode([
            'type' => 'action_result',
            'id' => $id,
            'success' => $success
        ]) . "\n";
        fwrite($this->socket, $payload);
    }

    public function close(): void {
        if ($this->socket) {
            fclose($this->socket);
            $this->socket = null;
        }
        if ($this->process) {
            proc_terminate($this->process);
            $this->process = null;
        }
    }
}

File: sdk/csharp/src/BdiClient.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Collections.Concurrent;
using System.Threading;
using System.Text.Json;
using System.Collections.Generic;

namespace Panteao.Sdk
{
    public class Panteao : IDisposable
    {
        private readonly TcpClient _client;
        private readonly NetworkStream _stream;
        private readonly StreamWriter _writer;
        private readonly StreamReader _reader;
        private readonly ConcurrentDictionary<string, Action<string[], Action<bool>>> _handlers;
        private readonly Thread _listenerThread;
        private bool _running;

        private readonly System.Diagnostics.Process _process;

        private static int GetFreePort()
        {
            var listener = new TcpListener(System.Net.IPAddress.Loopback, 0);
            listener.Start();
            int port = ((System.Net.IPEndPoint)listener.LocalEndpoint).Port;
            listener.Stop();
            return port;
        }

        private static string FindBinary()
        {
            bool isWin = System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows);
            string binName = isWin ? "panteao-engine.exe" : "panteao-engine";
            
            string baseDir = AppDomain.CurrentDomain.BaseDirectory;
            if (!string.IsNullOrEmpty(baseDir))
            {
                string cand1 = Path.Combine(baseDir, binName);
                if (File.Exists(cand1)) return cand1;
                string cand2 = Path.Combine(baseDir, "bin", binName);
                if (File.Exists(cand2)) return cand2;
            }
            
            string cwd = Directory.GetCurrentDirectory();
            string cand3 = Path.Combine(cwd, binName);
            if (File.Exists(cand3)) return cand3;
            string cand4 = Path.Combine(cwd, "bin", binName);
            if (File.Exists(cand4)) return cand4;
            
            return binName;
        }

        private static void PrintLog(string line)
        {
            if (string.IsNullOrWhiteSpace(line)) return;
            var match = System.Text.RegularExpressions.Regex.Match(line, @"^\[(.*?)\]\s(.*)");
            if (match.Success)
            {
                var rawName = match.Groups[1].Value;
                var parts = rawName.Split('.');
                var name = parts[parts.Length - 1];
                Console.WriteLine($"\x1b[36m[{name}]\x1b[0m {match.Groups[2].Value}");
            }
            else
            {
                Console.WriteLine($"\x1b[36m[MAS]\x1b[0m {line}");
            }
        }

        private static void DownloadEngine()
        {
            string osName = System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows) ? "win32" :
                            System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.OSX) ? "darwin" : "linux";
            string arch = System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture == System.Runtime.InteropServices.Architecture.Arm64 ? "arm64" : "x64";
            
            string sdkVersion = "1.1.17"; // To be bumped by CI
            string pkgName = $"panteao-engine-{osName}-{arch}";
            string url = $"https://registry.npmjs.org/{pkgName}/-/{pkgName}-{sdkVersion}.tgz";
            
            Console.WriteLine($"[Panteao] Downloading native engine for {osName}-{arch} (v{sdkVersion})...");
            
            string tempTgz = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "panteao.tgz");
            
            using (var client = new System.Net.Http.HttpClient())
            {
                var bytes = client.GetByteArrayAsync(url).GetAwaiter().GetResult();
                File.WriteAllBytes(tempTgz, bytes);
            }

            var p = new System.Diagnostics.Process();
            p.StartInfo.FileName = "tar";
            if (osName == "win32")
            {
                p.StartInfo.Arguments = $"-xf panteao.tgz package/bin/panteao-engine.exe";
                p.StartInfo.WorkingDirectory = AppDomain.CurrentDomain.BaseDirectory;
            }
            else
            {
                p.StartInfo.Arguments = $"-xz --strip-components=2 package/bin/panteao-engine";
                p.StartInfo.WorkingDirectory = AppDomain.CurrentDomain.BaseDirectory;
            }
            p.StartInfo.UseShellExecute = false;
            p.StartInfo.CreateNoWindow = true;
            p.Start();
            p.WaitForExit();
            
            if (osName == "win32")
            {
                string src = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "package", "bin", "panteao-engine.exe");
                string dest = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "panteao-engine.exe");
                if (File.Exists(src))
                {
                    if (File.Exists(dest)) File.Delete(dest);
                    File.Move(src, dest);
                    Directory.Delete(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "package"), true);
                }
            }
            else
            {
                string dest = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "panteao-engine");
                var chmod = new System.Diagnostics.Process();
                chmod.StartInfo.FileName = "chmod";
                chmod.StartInfo.Arguments = $"+x {dest}";
                chmod.Start();
                chmod.WaitForExit();
            }
            
            File.Delete(tempTgz);
            Console.WriteLine("[Panteao] Engine downloaded successfully.");
        }

        public Panteao(string host, int port, string project = null)
        {
            if (string.IsNullOrEmpty(host))
            {
                host = "127.0.0.1";
            }
            if (!string.IsNullOrEmpty(project))
            {
                if (port == 0)
                {
                    port = GetFreePort();
                }
                string bin = FindBinary();
                if (bin == "panteao-engine" || bin == "panteao-engine.exe")
                {
                    DownloadEngine();
                    bin = FindBinary();
                }
                
                _process = new System.Diagnostics.Process();
                _process.StartInfo.FileName = bin;
                _process.StartInfo.Arguments = $"\"{project}\" --port {port}";
                _process.StartInfo.UseShellExecute = false;
                _process.StartInfo.CreateNoWindow = true;
                _process.StartInfo.RedirectStandardOutput = true;
                _process.StartInfo.RedirectStandardError = true;
                _process.OutputDataReceived += (s, e) => PrintLog(e.Data);
                _process.ErrorDataReceived += (s, e) => PrintLog(e.Data);
                _process.Start();
                _process.BeginOutputReadLine();
                _process.BeginErrorReadLine();
                Thread.Sleep(800);
            }
            else
            {
                _process = null;
                if (port == 0) port = 44444;
            }

            _client = new TcpClient(host, port);
            _stream = _client.GetStream();
            _writer = new StreamWriter(_stream, new UTF8Encoding(false)) { AutoFlush = true };
            _reader = new StreamReader(_stream, Encoding.UTF8);

            while (true)
            {
                string line = _reader.ReadLine();
                if (line == null) throw new IOException("Connection lost during handshake");
                if (line.Contains("\"type\":\"mas_ready\""))
                {
                    break;
                }
            }

            _handlers = new ConcurrentDictionary<string, Action<string[], Action<bool>>>();
            _running = true;

            _listenerThread = new Thread(Listen) { IsBackground = true };
            _listenerThread.Start();
        }

        public System.Threading.Tasks.Task ConnectAsync()
        {
            return System.Threading.Tasks.Task.CompletedTask;
        }

        public void Wait()
        {
            while (_running)
            {
                Thread.Sleep(500);
            }
        }

        public void SendMsg(string performative, string sender, string receiver, string content)
        {
            var payload = $"{{\"type\":\"message\",\"performative\":\"{performative}\",\"sender\":\"{sender}\",\"receiver\":\"{receiver}\",\"content\":\"{content}\"}}";
            lock (_writer)
            {
                _writer.WriteLine(payload);
            }
        }

        public void Close()
        {
            Dispose();
        }

        public void SendPerception(string action, string perception)
        {
            var payload = $"{{\"type\":\"perception\",\"action\":\"{action}\",\"perception\":\"{perception}\"}}";
            lock (_writer)
            {
                _writer.WriteLine(payload);
            }
        }

        public void RegisterAction(string actionName, Action<string[], Action<bool>> callback)
        {
            _handlers[actionName] = callback;
        }

        private void SendActionResult(string id, bool success)
        {
            var payload = $"{{\"type\":\"action_result\",\"id\":\"{id}\",\"success\":{(success ? "true" : "false")}}}";
            lock (_writer)
            {
                _writer.WriteLine(payload);
            }
        }

        private void Listen()
        {
            try
            {
                while (_running)
                {
                    string line = _reader.ReadLine();
                    if (line == null) break;
                    HandleLine(line.Trim());
                }
            }
            catch { }
        }

        private void HandleLine(string line)
        {
            if (string.IsNullOrEmpty(line)) return;
            try
            {
                using var doc = JsonDocument.Parse(line);
                var root = doc.RootElement;
                if (root.TryGetProperty("type", out var typeProp) && typeProp.GetString() == "action")
                {
                    string id = root.GetProperty("id").GetString();
                    string rawAction = root.GetProperty("action").GetString();
                    var (name, args) = ParseAction(rawAction);

                    if (_handlers.TryGetValue(name, out var handler))
                    {
                        handler(args, (success) => SendActionResult(id, success));
                    }
                    else
                    {
                        SendActionResult(id, true);
                    }
                }
            }
            catch { }
        }

        private (string, string[]) ParseAction(string actionStr)
        {
            int parenIdx = actionStr.IndexOf('(');
            if (parenIdx == -1)
            {
                return (actionStr.Trim(), Array.Empty<string>());
            }
            string name = actionStr.Substring(0, parenIdx).Trim();
            string argsStr = actionStr.Substring(parenIdx + 1, actionStr.LastIndexOf(')') - parenIdx - 1);
            
            var args = new List<string>();
            var current = new StringBuilder();
            bool insideQuotes = false;
            int depthBrackets = 0;
            int depthParens = 0;
            foreach (char c in argsStr)
            {
                if (c == '"')
                {
                    insideQuotes = !insideQuotes;
                    current.Append(c);
                }
                else if (!insideQuotes && c == '[')
                {
                    depthBrackets++;
                    current.Append(c);
                }
                else if (!insideQuotes && c == ']')
                {
                    depthBrackets--;
                    current.Append(c);
                }
                else if (!insideQuotes && c == '(')
                {
                    depthParens++;
                    current.Append(c);
                }
                else if (!insideQuotes && c == ')')
                {
                    depthParens--;
                    current.Append(c);
                }
                else if (c == ',' && !insideQuotes && depthBrackets == 0 && depthParens == 0)
                {
                    args.Add(CleanArg(current.ToString()));
                    current.Clear();
                }
                else
                {
                    current.Append(c);
                }
            }
            if (current.Length > 0)
            {
                args.Add(CleanArg(current.ToString()));
            }
            return (name, args.ToArray());
        }

        private string CleanArg(string arg)
        {
            string s = arg.Trim();
            if (s.StartsWith("\"") && s.EndsWith("\"") && s.Length >= 2)
            {
                return s.Substring(1, s.Length - 2);
            }
            return s;
        }

        public void Dispose()
        {
            _running = false;
            try { _reader.Dispose(); } catch {}
            try { _writer.Dispose(); } catch {}
            try { _stream.Dispose(); } catch {}
            try { _client.Dispose(); } catch {}
            try
            {
                if (_process != null && !_process.HasExited)
                {
                    _process.Kill();
                    _process.Dispose();
                }
            }
            catch {}
        }
    }
}
